File: dotnet/lesson-1/Retriever.cs
// Retrieval engine: BM25 Okapi by hand (k1=1.5, b=0.75 — rank_bm25 defaults).
// Mirrors localrag/retriever.py's Bm25Retriever and build_retriever. Embeddings
// are out of scope for this C# port (see build_retriever's fallback notice).

using System.Text.RegularExpressions;

namespace LocalRag;

public interface IRetriever
{
    string Name { get; }
    List<Chunk> Search(string query, int k);
}

public partial class Bm25Retriever : IRetriever
{
    private const double K1 = 1.5;
    private const double B = 0.75;

    public string Name => "bm25";

    private readonly List<Chunk> _chunks;
    private readonly List<List<string>> _corpus;
    private readonly double[] _docLengths;
    private readonly double _avgDocLength;
    private readonly Dictionary<string, double> _idf;

    [GeneratedRegex("[a-z0-9]+")]
    private static partial Regex TokenRegex();

    public static List<string> Tokenize(string text) =>
        TokenRegex().Matches(text.ToLowerInvariant()).Select(m => m.Value).ToList();

    public Bm25Retriever(List<Chunk> chunks)
    {
        _chunks = chunks;
        _corpus = chunks.Count > 0
            ? chunks.Select(c => Tokenize(c.Text)).ToList()
            : new List<List<string>> { new() { "" } };

        _docLengths = _corpus.Select(d => (double)d.Count).ToArray();
        _avgDocLength = _docLengths.Length > 0 ? _docLengths.Average() : 0.0;

        // Document frequency per term, then the rank_bm25 Okapi IDF formula:
        // idf = log((N - df + 0.5) / (df + 0.5)). (Can be negative on tiny corpora.)
        var df = new Dictionary<string, int>();
        foreach (var doc in _corpus)
        {
            foreach (var term in doc.Distinct())
            {
                df[term] = df.GetValueOrDefault(term) + 1;
            }
        }
        var n = _corpus.Count;
        _idf = new Dictionary<string, double>(df.Count);
        foreach (var (term, freq) in df)
        {
            _idf[term] = Math.Log((n - freq + 0.5) / (freq + 0.5));
        }
    }

    private double[] GetScores(List<string> query)
    {
        var scores = new double[_corpus.Count];
        for (var i = 0; i < _corpus.Count; i++)
        {
            var doc = _corpus[i];
            var freqs = new Dictionary<string, int>();
            foreach (var term in doc)
            {
                freqs[term] = freqs.GetValueOrDefault(term) + 1;
            }
            var dl = _docLengths[i];
            double score = 0.0;
            foreach (var term in query)
            {
                if (!freqs.TryGetValue(term, out var tf) || !_idf.TryGetValue(term, out var idf))
                {
                    continue;
                }
                var denom = tf + K1 * (1 - B + B * dl / _avgDocLength);
                score += idf * (tf * (K1 + 1)) / denom;
            }
            scores[i] = score;
        }
        return scores;
    }

    public List<Chunk> Search(string query, int k)
    {
        if (_chunks.Count == 0)
        {
            return new List<Chunk>();
        }
        var scores = GetScores(Tokenize(query));
        // Stable descending sort by score (mirrors Python's sorted by key, reverse).
        var ranked = Enumerable.Range(0, _chunks.Count)
            .OrderByDescending(i => scores[i])
            .ToList();

        var top = ranked.Take(k).ToList();
        var best = scores[top[0]];
        if (best <= 0)
        {
            return top.Select(i => _chunks[i]).ToList();
        }
        return top.Where(i => scores[i] > 0).Select(i => _chunks[i]).ToList();
    }
}

public static class RetrieverFactory
{
    /// <summary>Pick a retriever from config. Embeddings fall back to BM25 in this port.</summary>
    public static IRetriever BuildRetriever(List<Chunk> chunks, Config config)
    {
        if (config.Retriever == "embeddings")
        {
            Console.WriteLine("[localrag] Embeddings are not ported in C# yet. Falling back to BM25.");
        }
        return new Bm25Retriever(chunks);
    }
}

File: node/lesson-1/src/retriever.js
// Retrieval engine: BM25 (default, zero setup).
//
// Mirrors localrag/retriever.py's Bm25Retriever, with BM25 Okapi implemented by
// hand to match rank_bm25's BM25Okapi defaults (k1=1.5, b=0.75, and the same
// `log((N - n + 0.5) / (n + 0.5))` IDF that can go negative on tiny corpora).
//
// PARITY NOTE: the EmbeddingRetriever is out of scope for this Node port. If
// `config.retriever === "embeddings"`, buildRetriever prints a one-line notice
// and falls back to BM25, like the Python does on embedding failure.

function tokenize(text) {
  return (text.toLowerCase().match(/[a-z0-9]+/g)) || [];
}

class Bm25Retriever {
  constructor(chunks) {
    this.name = "bm25";
    this.chunks = chunks;
    this.k1 = 1.5;
    this.b = 0.75;

    const corpus = chunks.length ? chunks.map((c) => tokenize(c.text)) : [[""]];
    this.corpusSize = corpus.length;
    this.docLengths = corpus.map((doc) => doc.length);
    const totalLen = this.docLengths.reduce((s, l) => s + l, 0);
    this.avgdl = totalLen / this.corpusSize;

    // Per-document term frequencies + document frequency per term.
    this.docFreqs = [];
    const df = new Map();
    for (const doc of corpus) {
      const freqs = new Map();
      for (const term of doc) freqs.set(term, (freqs.get(term) || 0) + 1);
      this.docFreqs.push(freqs);
      for (const term of freqs.keys()) df.set(term, (df.get(term) || 0) + 1);
    }

    // IDF, matching rank_bm25 BM25Okapi: log((N - n + 0.5) / (n + 0.5)).
    this.idf = new Map();
    for (const [term, n] of df.entries()) {
      this.idf.set(term, Math.log((this.corpusSize - n + 0.5) / (n + 0.5)));
    }
  }

  scores(queryTokens) {
    const out = new Array(this.corpusSize).fill(0);
    for (const term of queryTokens) {
      const idf = this.idf.get(term);
      if (idf === undefined) continue;
      for (let i = 0; i < this.corpusSize; i++) {
        const f = this.docFreqs[i].get(term);
        if (!f) continue;
        const denom = f + this.k1 * (1 - this.b + (this.b * this.docLengths[i]) / this.avgdl);
        out[i] += idf * ((f * (this.k1 + 1)) / denom);
      }
    }
    return out;
  }

  search(query, k) {
    if (!this.chunks.length) return [];
    const scores = this.scores(tokenize(query));
    const ranked = scores
      .map((_, i) => i)
      .sort((a, b) => scores[b] - scores[a]);
    // Return the top-k candidates by score and let the grounding prompt judge
    // relevance. (BM25 IDF can be negative on tiny corpora, so an absolute score
    // cutoff would wrongly discard everything.) Only drop the long tail that
    // scores strictly worse than the best, when there is a real signal.
    const top = ranked.slice(0, k);
    const best = scores[top[0]];
    if (best <= 0) {
      return top.map((i) => this.chunks[i]);
    }
    return top.filter((i) => scores[i] > 0).map((i) => this.chunks[i]);
  }
}

export function buildRetriever(chunks, config) {
  // Pick a retriever from config. Embeddings are not ported in Node; fall back
  // to BM25 with a clear message, mirroring the Python "never dead-end" design.
  if (config.retriever === "embeddings") {
    console.log(
      "[localrag] Embeddings not ported in Node. Falling back to BM25 (use the Python reference for embeddings)."
    );
  }
  return new Bm25Retriever(chunks);
}
